File: task18/main.py
# Вариант с файлом:
# with open("input.txt") as f:
#     lines = f.read().splitlines()
#     for line in lines:
#         parts = line.split()
#         if parts[0] == "insert": ht.insert(int(parts[1]), parts[2])
#         elif parts[0] == "search": print(ht.search(int(parts[1])))
#         elif parts[0] == "delete": ht.delete(int(parts[1]))

class HashTable:
    def __init__(self, size=10):
        self.size = size
        self.table = [[] for _ in range(size)]

    def _hash(self, key):
        return key % self.size

    def insert(self, key, value):
        idx = self._hash(key)
        for pair in self.table[idx]:
            if pair[0] == key:
                pair[1] = value
                return
        self.table[idx].append([key, value])

    def search(self, key):
        idx = self._hash(key)
        for pair in self.table[idx]:
            if pair[0] == key:
                return pair[1]
        return None

    def delete(self, key):
        idx = self._hash(key)
        self.table[idx] = [p for p in self.table[idx] if p[0] != key]

ht = HashTable()
while True:
    cmd = input("Команда (insert/search/delete/exit): ").strip()
    if cmd == "exit":
        break
    elif cmd == "insert":
        key = int(input("Ключ (число): "))
        value = input("Значение: ")
        ht.insert(key, value)
    elif cmd == "search":
        key = int(input("Ключ (число): "))
        print(ht.search(key))
    elif cmd == "delete":
        key = int(input("Ключ (число): "))
        ht.delete(key)

File: task18/main.cpp
#include <iostream>
#include <vector>
#include <list>
#include <string>
// Вариант с файлом:
// #include <fstream>
using namespace std;

class HashTable {
    int size;
    vector<list<pair<int, string>>> table;

    int hash(int key) {
        return key % size;
    }

public:
    HashTable(int size = 10) : size(size), table(size) {}

    void insert(int key, string value) {
        int idx = hash(key);
        for (auto& pair : table[idx]) {
            if (pair.first == key) {
                pair.second = value;
                return;
            }
        }
        table[idx].push_back({key, value});
    }

    string search(int key) {
        int idx = hash(key);
        for (auto& pair : table[idx]) {
            if (pair.first == key)
                return pair.second;
        }
        return "Not found";
    }

    void remove(int key) {
        int idx = hash(key);
        table[idx].remove_if([key](pair<int,string>& p) {
            return p.first == key;
        });
    }
};

int main() {
    // Вариант с файлом:
    // ifstream f("input.txt");
    // string cmd;
    // while (f >> cmd) {
    //     if (cmd == "insert") { int k; string v; f >> k >> v; ht.insert(k, v); }
    //     else if (cmd == "search") { int k; f >> k; cout << ht.search(k) << endl; }
    //     else if (cmd == "delete") { int k; f >> k; ht.remove(k); }
    // }

    HashTable ht;
    string cmd;
    while (true) {
        cout << "Команда (insert/search/delete/exit): ";
        cin >> cmd;
        if (cmd == "exit") break;
        else if (cmd == "insert") {
            int key; string value;
            cout << "Ключ (число): "; cin >> key;
            cout << "Значение: "; cin >> value;
            ht.insert(key, value);
        } else if (cmd == "search") {
            int key;
            cout << "Ключ (число): "; cin >> key;
            cout << ht.search(key) << endl;
        } else if (cmd == "delete") {
            int key;
            cout << "Ключ (число): "; cin >> key;
            ht.remove(key);
        }
    }
    return 0;
}